Rename Azure NextGen to Azure-Native (#5351)
pulumi/docs: content/blog/full-coverage-of-azure-resources-with-azure-native/index.md

---
title: "Full Coverage of Azure Resources with Azure-Native"

> date: 2021-02-24

Reviewer: ```suggestion
date: 2021-03-02
```

meta_desc: "Pulumi’s Azure-Native provider is the most comprehensive IaC solution for Microsoft Azure"
meta_image: azure-native.png
authors:
  - mikhail-shilkov
tags:
  - azure
---
 
Last September, we [announced the beta release of Pulumi Azure NextGen]({{< relref "/blog/announcing-nextgen-azure-provider" >}}): a new Microsoft Azure provider for Pulumi that combines same-day access to the entire [Azure API surface](https://docs.microsoft.com/en-us/rest/api/azure/) with the excellent Pulumi experience you know and love, including version-less resources, auto-naming, and auto-location.

Today, we’re excited to announce that this new provider is now the default way to manage Azure resources with Pulumi. We’re also excited to announce its final name: the native Azure provider for Pulumi, or “Azure-Native” for short. You can get started with the new provider using our newly-updated [getting started guide]({{< relref "/docs/get-started/azure" >}}).

<!--more-->

> Already use the [classic Pulumi-Azure provider](https://www.pulumi.com/docs/reference/pkg/azure/)? You can migrate to Azure-Native now using the [migration guide]({{< relref "/docs/intro/cloud-providers/azure/from-classic" >}}). If you prefer to wait until general availability, stay tuned for updates: we expect to reach GA in Spring of this year.

Reviewer: ```suggestion
Already use the [classic Pulumi-Azure provider](https://www.pulumi.com/docs/reference/pkg/azure/)? You can migrate to Azure-Native now using the [migration guide]({{< relref "/docs/intro/cloud-providers/azure/from-classic" >}}). If you prefer to wait until general availability, stay tuned for updates: we expect to reach general availability in Spring of this year.
```

Alongside this announcement, we’re also formalizing a new concept: a Pulumi “native provider”. A native provider provides functionality mapped directly from the underlying API; in the case of the new Azure Provider, functionality is mapped directly from the Azure Resource Manager API surface. Going forward, you’ll be able to find these next-generation Pulumi providers directly from the name: just look for the “-Native” suffix.

## Native to Azure

> The native Azure provider implements the best support possible for the Azure platform in Pulumi. We designed the native provider to expose the entire API surface of Azure to developers and operators, now and forever.
> 
> As of today, the native provider supports 1010 resource types, 120 of them added since the initial preview in September 2020. Every property of each resource is always represented in the SDKs.
> 
> Unlike the classic Azure provider, which requires manual work to keep updated, we designed the native provider to stay always up-to-date with Azure API additions and changes. We generate Pulumi SDKs for the native Azure provider automatically from Azure API specifications published by Microsoft. We publish daily builds of the provider and have published 210 versions of the provider in the last six months.

Reviewer: ```suggestion
The native Azure provider implements the best possible support for the Azure platform in Pulumi. It exposes the entire API surface of Azure to developers and operators, now and forever. As of today, the native provider supports 1,010 resource types, 120 of them added since the initial preview in September 2020. Every property of each resource is always represented in the SDKs.

Unlike the classic Azure provider, which requires manual work to keep updated, we designed the native provider to stay always up-to-date with Azure API additions and changes. We generate Pulumi SDKs for the native Azure provider automatically from Azure API specifications published by Microsoft. We publish daily builds of the provider and have published 210 versions of the provider in the last six months.
```

## Native to Pulumi

The native Azure provider comes with several significant improvements over the Azure NextGen preview. These changes bring more features of the Pulumi experience to the native Azure provider.

> ### Top-level resources

Reviewer 2: ```suggestion
### Top-level Resources
```

{
   public MyStack()
   {
       var resourceGroup = new ResourceGroup("resourceGroup");

       var storageAccount = new StorageAccount("sa", new StorageAccountArgs
       {
           ResourceGroupName = resourceGroup.Name,
           Sku = new SkuArgs
           {
               Name = SkuName.Standard_LRS
           },
           Kind = Kind.StorageV2
       });
   }
}
```

{{< /choosable >}}
{{< choosable language go >}}

```go
package main

import (
   "github.com/pulumi/pulumi-azure-native/sdk/go/azure/resources"
   "github.com/pulumi/pulumi-azure-native/sdk/go/azure/storage"
   "github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
   pulumi.Run(func(ctx *pulumi.Context) error {
       resourceGroup, err := resources.NewResourceGroup(ctx, "resourceGroup")
       if err != nil {
           return err
       }

       account, err := storage.NewStorageAccount(ctx, "sa", &storage.StorageAccountArgs{
           ResourceGroupName: resourceGroup.Name,
           Sku: &storage.SkuArgs{
               Name: storage.SkuName_Standard_LRS,
           },
           Kind: storage.KindStorageV2,
       })

       return err
   })
}
```

{{< /choosable >}}
{{< /chooser >}}

> The top-level resources will use stable API versions and take no arbitrary breaking changes within a given major version of the package (minor version while in preview).

Reviewer: ```suggestion
The top-level resources will use stable API versions and take no arbitrary breaking changes within a given major version of the package once we reach general availability; during the preview, breaking changes may appear in `0.x` minor versions).
```

---
title: "Full Coverage of Azure Resources with Azure-Native"
date: 2021-03-02
meta_desc: "Pulumi’s Azure-Native provider is the most comprehensive IaC solution for Microsoft Azure"
meta_image: azure-native.png
authors:
  - mikhail-shilkov
tags:
  - azure
---
 
Last September, we [announced the beta release of Pulumi Azure NextGen]({{< relref "/blog/announcing-nextgen-azure-provider" >}}): a new Microsoft Azure provider for Pulumi that combines same-day access to the entire [Azure API surface](https://docs.microsoft.com/en-us/rest/api/azure/) with the excellent Pulumi experience you know and love, including version-less resources, auto-naming, and auto-location.

Today, we’re excited to announce that this new provider is now the default way to manage Azure resources with Pulumi. We’re also excited to announce its final name: the native Azure provider for Pulumi, or “Azure-Native” for short. You can get started with the new provider using our newly-updated [getting started guide]({{< relref "/docs/get-started/azure" >}}).

<!--more-->

> Already use the [classic Pulumi-Azure provider](https://www.pulumi.com/docs/reference/pkg/azure/)? You can migrate to Azure-Native now using the [migration guide]({{< relref "/docs/intro/cloud-providers/azure/from-classic" >}}). If you prefer to wait until general availability, stay tuned for updates: we expect to reach general availability in Spring of this year.

Reviewer 3: ```suggestion
Already use the [classic Pulumi-Azure provider](https://www.pulumi.com/docs/reference/pkg/azure/)? You can migrate to Azure-Native now using the [migration guide]({{< relref "/docs/intro/cloud-providers/azure/from-classic" >}}). If you prefer to wait until general availability, stay tuned for updates: we expect to reach general availability in the spring of this year.
```

> Alongside this announcement, we’re also formalizing a new concept: a Pulumi “native provider”. A native provider provides functionality mapped directly from the underlying API; in the case of the new Azure Provider, functionality is mapped directly from the Azure Resource Manager API surface. Going forward, you’ll be able to find these next-generation Pulumi providers directly from the name: just look for the “-Native” suffix.

Reviewer 3: ```suggestion
Alongside this announcement, we’re also formalizing a new concept: a Pulumi “native provider.” A native provider provides functionality mapped directly from the underlying API; in the new Azure Provider, functionality is mapped directly from the Azure Resource Manager API surface. Going forward, you’ll be able to find these next-generation Pulumi providers directly from the name by looking for the “-Native” suffix.
```

           {
               Name = SkuName.Standard_LRS
           },
           Kind = Kind.StorageV2
       });
   }
}
```

{{< /choosable >}}
{{< choosable language go >}}

```go
package main

import (
   "github.com/pulumi/pulumi-azure-native/sdk/go/azure/resources"
   "github.com/pulumi/pulumi-azure-native/sdk/go/azure/storage"
   "github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
   pulumi.Run(func(ctx *pulumi.Context) error {
       resourceGroup, err := resources.NewResourceGroup(ctx, "resourceGroup", nil)
       if err != nil {
           return err
       }

       account, err := storage.NewStorageAccount(ctx, "sa", &storage.StorageAccountArgs{
           ResourceGroupName: resourceGroup.Name,
           Sku: &storage.SkuArgs{
               Name: storage.SkuName_Standard_LRS,
           },
           Kind: storage.KindStorageV2,
       })

       return err
   })
}
```

{{< /choosable >}}
{{< /chooser >}}

The top-level resources will use stable API versions and take no arbitrary breaking changes within a given major version of the package once we reach general availability; during the preview, breaking changes may appear in `0.x` minor versions.

> We still provide access to every version of every ARM API if you want to use a brand-new version or a specific version from the past. Each API version is mapped to its own module or namespace within the resource provider.

Reviewer 3: ```suggestion
We still provide access to every version of every ARM API if you want to use the latest version or a specific version from the past. Each API version is mapped to its own module or namespace within the resource provider.
```

{{< chooser language "typescript,python,csharp,go" >}}

{{< choosable language typescript >}}

```typescript
import * as azure from "@pulumi/azure-native";
const resourceGroup = new azure.resources.ResourceGroup("rg");
```

{{< /choosable >}}
{{< choosable language python >}}

```python
import pulumi_azure_native as azure
resource_group = azure.resources.ResourceGroup('rg')
```

{{< /choosable >}}
{{< choosable language csharp >}}

```csharp
using Pulumi;
using Pulumi.AzureNative.Resources;
class MyStack : Stack
{
   public MyStack()
   {
       var resourceGroup = new ResourceGroup("rg");
   }
}
```

{{< /choosable >}}
{{< choosable language go >}}

```go
package main

import (
   resources "github.com/pulumi/pulumi-azure-native/sdk/go/azure/resources"
   "github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
   pulumi.Run(func(ctx *pulumi.Context) error {
       _, err := resources.NewResourceGroup(ctx, "resourceGroup", nil)
       return err
   })
}
```

{{< /choosable >}}
{{< /chooser >}}

Pass an explicit value to the `resourceGroupName` property if you want to control the resource's exact name.

> However, suppose you create a sub-resource, e.g., a container under a storage account, a database under a database account, or a slot under a web app. In that case, Pulumi will not append any random suffix because the names of those resources don’t have to be globally unique.

Reviewer 3: ```suggestion
However, suppose you create a sub-resource, e.g., a container under a storage account, a database under a database account, or a slot under a web app. In that case, Pulumi will not append a random suffix because those resources names don’t have to be globally unique.
```

Migration from Azure NextGen to Azure-Native is simple and non-disruptive:

- Reference the [`pulumi-azure-native`](https://github.com/pulumi/pulumi-azure-native) package from your program.
- Update all import/using statements from `nextgen` to `native`.
- Run `pulumi up` and make sure the program ran successfully. The preview should show no changes.
- If so, choose `yes` to run the update that will overwrite the state to the native provider.

Note that the Azure resources are not affected by this migration.

## Getting Started

If you are new to Pulumi, follow our [Get Started]({{< relref "/docs/get-started/azure" >}}) guide.

From now on, the `azure-*` templates will reference the native Azure provider. It's easier to get started than ever:

{{< chooser language "typescript,python,csharp,go" >}}

{{< choosable language typescript >}}

```sh
$ pulumi new azure-typescript
```

{{< /choosable >}}
{{< choosable language python >}}

```sh
$ pulumi new azure-python
```

{{< /choosable >}}
{{< choosable language csharp >}}

```sh
$ pulumi new azure-csharp
```

{{< /choosable >}}
{{< choosable language go >}}

```sh
$ pulumi new azure-go
```

{{< /choosable >}}
{{< /chooser >}}

> Several larger examples are available in the Pulumi Examples repo:

Reviewer 3: ```suggestion
Several more extensive examples are available in the Pulumi Examples repo:
```
